### ISSUE TITLE
vim.treesitter query problem

### ISSUE THREAD

**Issue by sergii4** [neovim/neovim#22055]
### Describe the bug

`vim.treesitter` can't capture big number of nodes of one pattern. Fox example, if you have a Go test with more than 5 tables test cases, vim.treesitter starts skipping previous tests names. That query highlights 5 test cases name instead of 6
<img width="1915" alt="Screenshot 2023-01-30 at 12 32 16" src="https://user-images.githubusercontent.com/13333099/215478058-388af988-01f7-4086-8f47-de253b680d14.png">

It seems that problem somewhere in `vim.treesitter.parse_query` and iterating over it with `:iter_matches`

### Steps to reproduce

**With nvim treesitter playground:**

- [Install](https://github.com/nvim-treesitter/playground) nvim treesitter with playgound
-  Create a Go test file, for example, `table_test.go`:
```
package main

import (
	"fmt"
	"net/http"
	"testing"
)

func TestSomeTest(t *testing.T) {
	tt := []struct {
		name   string
		method string
		url    string
		apiKey string
		status int
	}{
		{name: "AccessDenied1", method: http.MethodGet, url: "/api/nothing", apiKey: "lalala", status: http.StatusForbidden},
		{name: "AccessDenied2", method: http.MethodGet, url: "/api/nothing", apiKey: "lalala", status: http.StatusForbidden},
		{name: "AccessDenied3", method: http.MethodGet, url: "/api/nothing", apiKey: "lalala", status: http.StatusForbidden},
		{name: "AccessDenied4", method: http.MethodGet, url: "/api/nothing", apiKey: "lalala", status: http.StatusForbidden},
		{name: "AccessDenied5", method: http.MethodGet, url: "/api/nothing", apiKey: "lalala", status: http.StatusForbidden},
		{name: "AccessDenied6", method: http.MethodGet, url: "/api/nothing", apiKey: "lalala", status: http.StatusForbidden},
	}

	for _, tc := range tt {
		tc := tc
		t.Run(tc.name, func(_ *testing.T) {
			fmt.Println(tc.name, tc.method, tc.url, tc.apiKey, tc.status)
		})
	}

}
```
- open playground (`:TSPlaygroundToggle`), open query editor(`o`), use that query:
```
(block
      (short_var_declaration
        left: (expression_list
          (identifier) @test.cases)
        right: (expression_list
          (composite_literal
            body:(literal_value
              (literal_element
                (literal_value
                  (keyed_element
                    (literal_element
                      (identifier) @test.field.name1)
                    (literal_element
                      (interpreted_string_literal) @test.name)))) @test.definition))))
      (for_statement
        body: (block
          (call_expression
            function: (selector_expression
              field: (field_identifier) @test.method)
              (#match? @test.method "^Run$")
            arguments: (argument_list
              (selector_expression
                field: (field_identifier) @test.field.name2
                (#eq? @test.field.name1 @test.field.name2)
		))))))

```
- put cursor under capture `@test.name`

**With programming and output:**

- create `table_test.go` file with content as described above
- create `run.lua` file with content
```
local q = require"vim.treesitter.query"

local function i(value)
    print(vim.inspect(value))
end

local bufnr = 9 -- output Go test file buffer number with `:echo bufnr()`

local language_tree = vim.treesitter.get_parser(bufnr, 'go')
local syntax_tree = language_tree:parse()
local root = syntax_tree[1]:root()

local query = vim.treesitter.parse_query('go', [[
(block
      (short_var_declaration
        left: (expression_list
          (identifier) @test.cases)
        right: (expression_list
          (composite_literal
            body:(literal_value
              (literal_element
                (literal_value
                  (keyed_element
                    (literal_element
                      (identifier) @test.field.name1)
                    (literal_element
                      (interpreted_string_literal) @test.name)))) @test.definition))))
      (for_statement
        body: (block
          (call_expression
            function: (selector_expression
              field: (field_identifier) @test.method)
              (#match? @test.method "^Run$")
            arguments: (argument_list
              (selector_expression
                field: (field_identifier) @test.field.name2
                (#eq? @test.field.name1 @test.field.name2)
		))))))   

]])

for _, captures, metadata in query:iter_matches(root, bufnr) do
    i(q.get_node_text(captures[3], bufnr))
end
```
- split window having Go test in one part and lua in another
- execute lua file `:luafile %`

You will see and output such as
```
'"AccessDenied2"'
'"AccessDenied3"'
'"AccessDenied4"'
'"AccessDenied5"'
'"AccessDenied6"'
```

### Expected behavior

`vim.treesitter` captures each node successfully

### Neovim version (nvim -v)

NVIM v0.8.1

### Vim (not Nvim) behaves the same?

vim doesn't support tresitter natively

### Operating system/version

macOS 13.1 (22C65)

### Terminal name/version

kitty 0.26.5

### $TERM environment variable

tmux-256color

### Installation

homebrew

**Comment by theHamsta** [comment]
Likely a duplicate of https://github.com/nvim-treesitter/nvim-treesitter/issues/3977 (there should be also Neovim bugs related to this). I assume this is also due to the set match limit: https://github.com/neovim/neovim/blob/4a78ed60a3acd3a5694cc34debd4d91bed96195a/src/nvim/lua/treesitter.c#L1219

**Comment by sergii4** [comment]
Hi @theHamsta! Thanks for the response. Do I need to build neovim locally to check it up?
What I have found so far is that for some reason first match marked as [inactive](https://github.com/neovim/neovim/blob/21257d8580780a5e28f06562cea4f8889e6b0932/runtime/lua/vim/treesitter/query.lua#L633).

I copied and tweaked that code locally:
```
  local function iter()
    local pattern, match = raw_iter()
    local metadata = {}
    if match ~= nil then
   	  i(q.get_node_text(match[4], bufnr))
      local active = query:match_preds(match, pattern, source)
	  print('active', active)
      if not active then
        return iter() -- tail call: try next match
      end

      query:apply_directives(match, pattern, source, metadata)
    end
```
and get an output that threats node as inactive:
```
'{name: "AccessDenied1", method: http.MethodGet, url: "/api/nothing", apiKey: "lalala", status: http.StatusForbidden}'
active false
'{name: "AccessDenied2", method: http.MethodGet, url: "/api/nothing", apiKey: "lalala", status: http.StatusForbidden}'
active true
'"AccessDenied2"'
...
```

**Comment by clason** [comment]
But match limit was necessary to work around significant performance issues. Maybe it is necessary to make that configurable.

**Comment by sergii4** [comment]
Hi @clason. `128` work for that case, 6 test cases. It works even for 10 test cases. But it doesn't work for 20 test cases which requires setting to `256`. I think 20 test cases is upper limit for 95% golang tests, so `256` should be fine

**Comment by clason** [comment]
@theHamsta You had some performance benchmarks, or at least some problematic files to test with, in the linked issue -- can you test `128` and `256` and see if they behave acceptably?

**Comment by theHamsta** [comment]
If possible, I'd rather set a timeout for the query itself and raise the limit for the match limit unless the match limit is necessary for a spot of the code that's not canceleable. It would be much easier for users and us to specify a time limit than some abstract number. A timelimit could also be easier communicated to a user than unknowingly hitting a abstract limit and there is no way to recognize that in the result. A timeout could also be easier adapted to the usecase at hand.

**Comment by clason** [comment]
Sure, PR welcome!

**Comment by clason** [comment]
But still, if we can raise the default number even a bit without noticeable performance regression, that sounds like a freebie to me.

**Comment by sergii4** [comment]
> But still, if we can raise the default number even a bit without noticeable performance regression, that sounds like a freebie to me.

It would be nice to have some working defaults. It looks difficult to implement a plugin cause of really TS cursor's strict limit, which is quite easy to reach out. 

Timeout would be nice to have to polish development. But reasonable default can unblock most of current issues and reduce barrier of using TS

**Comment by clason** [comment]
For the record, Helix uses 256: https://github.com/helix-editor/helix/blob/ddc5bf4e606230dd8bc20c59b1cc114e93bbf1c0/helix-core/src/syntax.rs#L380

I'd be fine with bumping to that?
